LLK/scratch-gui
# Create cloud vars

### Resolves

Towards having fully functional cloud variables in 3.0.

### Proposed Changes

Adds the ability to create cloud variables in a 3.0 project.
- adds a cloud variable option to the variable modal
- refactors cloud data connection logic into a separate HOC. 

#### Known Issues/TODO
- There is still more logic to come (not in this PR) about checking whether the user is a scratcher and about handling the cases where the user is viewing someone else's cloud project in editor mode.
- Adding a tooltip for why the cloud option becomes disabled (e.g. notifying the user about the cloud variable limit)
- Adding alerts/info bubbles about cloud variable usage when creating the first cloud variable in a project (like in 2.0)


### Reason for Changes

Cloud variables!

### Test Coverage

Manual testing:

On a local scratch environment (api, scratchr2, www, projects, assets), create a new account (doesn't yet have to be a scratcher) and ensure the following:
- [ ] navigate to /preview/editor to create a new project, create a cloud variable using the variable modal, see that the cloud variable gets created
- [ ] You should not be able to create more than 8 cloud variables (cloud var option becomes disabled in menu)
- [ ] Create code to modify the cloud variable, save, and share the project.
- [ ] Create another account and interact with the cloud project and see that the value of the cloud variable updates on both accounts (use separate browsers)
- [ ] In one browser, sign out and try to interact with the project (the cloud variable should neither send updates so that they show up on other accounts, nor should they receive updates made from another account).
- [ ] In signed out mode, in the editor, the cloud variable option should not show up in the variable modal
- [ ] Renaming a cloud variable should append a cloud + space if not present at the front of the rename, and should not append an extra one if it is present
- [ ] Cloud variable option should not show up in the variable modal in standalone gui

### Related PRs
- LLK/scratch-vm#1755
- LLK/scratch-blocks#1782

### Browser Coverage
Check the OS/browser combinations tested (At least 2)

Mac
 
Windows
 
Chromebook
 
iPad

Android Tablet
